### PR TITLE
Optimisation: Optimise core api by changing the way concepts are built, resulting in less validation. 

### PR DESCRIPTION
**Pull request by fppt**

### CODE REVIEW

**Comment by fppt** [comment]
screen shot 2017-09-18 at 18 23 54

Top is before this PR and bottom is after. Looks like we gained almost 20 writes per second.

**Comment by fppt** [comment]
!rtg - ai.grakn.test.engine.postprocessing.PostProcessingIT.checkThatDuplicateResourcesAtLargerScaleAreMerged

**Comment by fppt** [comment]
!rtg - ai.grakn.test.engine.postprocessing.PostProcessingIT.checkThatDuplicateResourcesAtLargerScaleAreMerged

File: grakn-core/src/main/java/ai/grakn/exception/PropertyNotUniqueException.java [hunk 1]
```diff
@@ -55,10 +56,10 @@ public static PropertyNotUniqueException cannotChangeProperty(Element mutatingCo
     }
 
     /**
-     * Thrown when trying to create a concept using a unique property which is already taken.
+     * Thrown when trying to create a {@link SchemaConcept} using a unique property which is already taken.
```
Comment (aelred) [inline grakn-core/src/main/java/ai/grakn/exception/PropertyNotUniqueException.java]: Should be `{@link Concept}`

File: grakn-kb/src/main/java/ai/grakn/kb/internal/GraknTxAbstract.java [hunk 2]
```diff
@@ -429,13 +415,39 @@ public EntityType putEntityType(Label label) {
...
+        } else if (!baseType.equals(schemaConcept.baseType())) {
+            throw labelTaken(schemaConcept);
+        }
+
+        //Check if the type we got is correct
+        SchemaConceptImpl finalSchemaConcept = schemaConcept;
         T finalType = validateSchemaConcept(schemaConcept, baseType, () -> {
-            if (Schema.MetaSchema.isMetaLabel(label)) throw GraknTxOperationException.reservedLabel(label);
-            throw PropertyNotUniqueException.cannotCreateProperty(schemaConcept, Schema.VertexProperty.SCHEMA_LABEL, label);
+            throw labelTaken(finalSchemaConcept);
```
Comment (aelred) [inline grakn-kb/src/main/java/ai/grakn/kb/internal/GraknTxAbstract.java]: Another optimisation: maybe we don't need to check this if the schema concept was just created.

Comment (fppt) [reply grakn-kb/src/main/java/ai/grakn/kb/internal/GraknTxAbstract.java]: Maybe, but the label is cached so we wont get much from this. Will check though.

**Comment by fppt** [comment]
@aelred re-review please.

File: grakn-kb/src/main/java/ai/grakn/kb/internal/GraknTxAbstract.java [hunk 3]
```diff
@@ -444,18 +443,13 @@ public EntityType putEntityType(Label label) {
...
-            throw labelTaken(finalSchemaConcept);
-        });
-
         //Automatic shard creation - If this type does not have a shard create one
-        if (!Schema.MetaSchema.isMetaLabel(label) && !SchemaConceptImpl.from(schemaConcept).vertex().getEdgesOfType(Direction.IN, Schema.EdgeLabel.SHARD).findAny().isPresent()) {
-            SchemaConceptImpl.from(schemaConcept).createShard();
-        }
+        //if (!Schema.MetaSchema.isMetaLabel(label) && !SchemaConceptImpl.from(schemaConcept).vertex().getEdgesOfType(Direction.IN, Schema.EdgeLabel.SHARD).findAny().isPresent()) {
+        //    SchemaConceptImpl.from(schemaConcept).createShard();
+        //}
```
Comment (aelred) [inline grakn-kb/src/main/java/ai/grakn/kb/internal/GraknTxAbstract.java]: 🐐

File: grakn-kb/src/main/java/ai/grakn/kb/internal/CommitLog.java [hunk 4]
```diff
@@ -52,11 +48,16 @@
 public class CommitLog {
     private final ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
     private final Map<ConceptId, Long> newInstanceCount = new ConcurrentHashMap<>();
-    private final Set<Attribute> newAttributes = ConcurrentHashMap.newKeySet();
+    private final Map<String, Set<String>> newAttributes = new ConcurrentHashMap<>();
```
Comment (aelred) [inline grakn-kb/src/main/java/ai/grakn/kb/internal/CommitLog.java]: for another time: these kinds of structures can use guava `Multimap`

**Comment by fppt** [comment]
@aelred re-review please

**Pull request merged**